src/stactools/nrcan_spot_ortho/utils.py

```diff
@@ -0,0 +1,158 @@
+from ftplib import error_perm
+import os
+from urllib.parse import urlparse
+from pystac import Link
+from pystac.stac_io import DefaultStacIO
+from typing import Union, Any
+import zipfile
+import logging
+from subprocess import Popen, PIPE, STDOUT
+import boto3
+from botocore.errorfactory import ClientError
+import glob
+
+logger = logging.getLogger(__name__)
+
+
+class CustomStacIO(DefaultStacIO):
+    def __init__(self):
+        self.s3 = boto3.resource("s3")
+
+    def read_text(self, source: Union[str, Link], *args: Any,
+                  **kwargs: Any) -> str:
+        parsed = urlparse(source)
+        if parsed.scheme == "s3":
+            bucket = parsed.netloc
+            key = parsed.path[1:]
+
+            obj = self.s3.Object(bucket, key)
+            return obj.get()["Body"].read().decode("utf-8")
+        else:
+            return super().read_text(source, *args, **kwargs)
+
+    def write_text(self, dest: Union[str, Link], txt: str, *args: Any,
+                   **kwargs: Any) -> None:
+        parsed = urlparse(dest)
+        if parsed.scheme == "s3":
+            bucket = parsed.netloc
+            key = parsed.path[1:]
+            s3 = boto3.resource("s3")
+            s3.Object(bucket, key).put(Body=txt, ContentEncoding="utf-8")
+        else:
+            super().write_text(dest, txt, *args, **kwargs)
+
+
+def call(command):
+    def log_subprocess_output(pipe):
+        for line in iter(pipe.readline, b''):  # b'\n'-separated lines
+            logger.info(line.decode("utf-8").strip('\n'))
+
+    process = Popen(command, stdout=PIPE, stderr=STDOUT)
+    with process.stdout:
+        log_subprocess_output(process.stdout)
+    return process.wait()  # 0 means success
+
+
+def upload_to_s3(parsed, local_path):
+    bucket = parsed.netloc
+    key = parsed.path[1:]
+    s3 = boto3.resource("s3")
+    print(f"Uploading {os.path.basename(local_path)}")
+    s3.Bucket(bucket).upload_file(local_path, key)
+
+
+def get_existing_paths(directory, ending):
+    parsed = urlparse(directory)
+
+    if parsed.scheme == "s3":
+        bucket = parsed.netloc
+        s3 = boto3.client('s3')
+        paginator = s3.get_paginator('list_objects_v2')
+        pages = paginator.paginate(Bucket=bucket) #, Prefix='prefix'
+        paths = []
+        for i, page in enumerate(pages):
+            print(f"S3 page {i+1}/?")
+            paths += [f"{directory}/{d['Key']}" for d in page['Contents'] 
+            if d['Key'][-len(ending):] == ending]
+        return paths
+
+    else:
+        return glob.glob(f"{directory}{os.sep}**{os.sep}*{ending}",
+                         recursive=True)
+
+
+def file_exists(path, paths_s3):
```

Comment:
There is a new `href_exists` function in `stactools` that does something similar, but this looks good too: https://github.com/stac-utils/stactools/blob/main/src/stactools/core/utils/__init__.py#L16-L23.

Reply:
Noted, I'll use that for future implementations. Turns out I don't even use this function after all so I'll remove it. I collect all existing paths upfront with `get_existing_paths` then compare with that list. Made s3 interactions faster!